File: src/Solutions/Euler336.hs
module Solutions.Euler336 where
import Data.List(delete)

-- All arrangements in lexigographical order
permutations :: Int -> [String]
permutations n = go (take n ['A'..]) where
    go [] = [[]]
    go l = [a:b | a <- l, b <- go (delete a l)]

splitOn :: Eq a => a -> [a] -> ([a], [a])
splitOn e [] = ([], [])
splitOn e (h:t) = if e == h then ([], h:t) else (h:p, q) where (p, q) = splitOn e t

countRotations :: String -> Int
countRotations = go 0 'A' where
    go k i [] = k
    go k i (h:t) = if h == i
        then go k (succ i) t
        else case splitOn i (h:t) of
            (l, [i]) -> go (k + 1) (succ i) (reverse l)
            (l1, l2) -> go (k + 2) (succ i) (tail $ reverse $ l1 ++ reverse l2)

findArr :: Int -> Int -> String
findArr whichArr carriages =
    (filter (\c -> countRotations c == 2 * carriages - 3) $ permutations carriages) !! (whichArr - 1)

euler336 :: IO String
euler336 = return $ findArr 2011 11

File: rust/src/solutions/euler336.rs
use itertools::Itertools;

fn permutations(carriages: u8) -> impl Iterator<Item = Vec<char>> {
    (0..carriages)
        .map(|d| ('A' as u8 + d) as char)
        .permutations(carriages as usize)
}

fn next(c: char) -> char {
    (c as u8 + 1) as char
}

fn count_rotations(arrangement: &Vec<char>) -> usize {
    fn count(i: char, arr: &mut Vec<char>) -> usize {
        if let Some(&h) = arr.first() {
            if h == i {
                arr.remove(0);
                count(next(i), arr)
            } else {
                let i_pos = arr.iter().position(|&c| c == i).unwrap_or(arr.len());
                if i_pos == arr.len() - 1 {
                    arr.pop();
                    arr.reverse();
                    1 + count(next(i), arr)
                } else {
                    arr[i_pos..].reverse();
                    arr.pop();
                    arr.reverse();
                    2 + count(next(i), arr)
                }
            }
        } else {
            0
        }
    }
    count('A', &mut arrangement.clone())
}

fn find_arrangement(n: usize, carriages: u8) -> Vec<char> {
    permutations(carriages)
        .filter(|arrangement| count_rotations(arrangement) == 2 * carriages as usize - 3)
        .nth(n - 1)
        .unwrap()
}

pub fn euler336() -> String {
    find_arrangement(2011, 11).into_iter().collect()
}
